src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -44,1115 +44,716 @@
-                                      joystickList.end(),
-                                      [](const JoystickRecord& joystickRecord) { return !joystickRecord.plugged; }),
-                       joystickList.end());
-
-    if (FAILED(result))
-    {
-        err() << "Failed to enumerate DirectInput devices: " << result << std::endl;
-
-        return;
-    }
-
-    // Assign unused joystick indices to devices that were newly connected
-    for (unsigned int i = 0; i < Joystick::Count; ++i)
-    {
-        for (JoystickRecord& record : joystickList)
+        if (targetJoystick.m_useXInput)
         {
-            if (record.index == i)
-                break;
-
-            if (record.index == Joystick::Count)
-            {
-                record.index = i;
-                break;
-            }
+            // we're getting the information from another API, return.
+            return;
         }
-    }
-}
 
+        uiResult = GetRawInputDeviceInfoW(deviceHandle, RIDI_PREPARSEDDATA, nullptr, &bufferSize);
+        if (uiResult == UINT_MAX)
+        {
+            // that didn't work!
+            lastError = GetLastError();
+            err() << "GetRawInputDeviceInfoW returned [" << uiResult << "] unexpectedly! GetLastError: [" << lastError
+                  << "]" << std::endl;
+            return;
+        }
 
-////////////////////////////////////////////////////////////
-bool JoystickImpl::openDInput(unsigned int index)
-{
-    // Initialize DirectInput members
-    m_device = nullptr;
+        uiResult = GetRawInputDeviceInfoW(deviceHandle, RIDI_PREPARSEDDATA, preparsedDataChunk.data(), &bufferSize);
+        if (uiResult == UINT_MAX)
+        {
+            // that didn't work!
+            lastError = GetLastError();
+            err() << "GetRawInputDeviceInfoW returned [" << uiResult << "] unexpectedly! GetLastError: [" << lastError
+                  << "]" << std::endl;
+            return;
+        }
 
-    m_axes.fill(-1);
-    m_buttons.fill(-1);
+        auto* preparsedData = reinterpret_cast<PHIDP_PREPARSED_DATA>(preparsedDataChunk.data());
 
-    m_deviceCaps        = {};
-    m_deviceCaps.dwSize = sizeof(DIDEVCAPS);
-    m_state             = JoystickState();
-    m_buffered          = false;
+        if (FAILED(hresult = HidP_GetCaps(preparsedData, &hCaps)))
+        {
+            // that didn't work!
+            lastError = GetLastError();
+            err() << "GetRawInputDeviceInfoW returned [" << hresult << "] unexpectedly! GetLastError: [" << lastError
+                  << "]" << std::endl;
+            return;
+        }
 
-    // Search for a joystick with the given index in the connected list
-    for (const JoystickRecord& record : joystickList)
-    {
-        if (record.index == index)
+        std::uint16_t    capsLength = hCaps.NumberInputButtonCaps;
+        HIDP_BUTTON_CAPS hButtonCaps{};
+        if (FAILED(hresult = HidP_GetButtonCaps(HidP_Input, &hButtonCaps, &capsLength, preparsedData)))
```

Comment:
```suggestion
        if (const auto result = HidP_GetButtonCaps(HidP_Input, &hButtonCaps, &capsLength, preparsedData); FAILED(result))
```
This ensures `result`'s scope is limited to this `if` statement instead of being visible to the larger scope. This applies to other places where `FAILED` is used so please check those as well.